fabtests: Add unit test for fi_endpoint interfaces

rajachan opened the pull request:
This is a simple unit test to test the fi_endpoint interfaces. It tests
opening/closing endpoints with the maximum number of EPs as advertised
by a provider, at random, and with different context sizes. It also
tests getting and setting EP opflags.

Signed-off-by: Raghu Raja <craghun@amazon.com>

rajachan commented:
Hey, the test works. The travis failure is from sockets possibly hitting a off-by-one error somewhere:

>    Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[128] Opened[129]
    FAIL: 
  
What's the Intel CI failure?

shefty commented:
```
name:   fi_ep_test -p "udp"
  result: Fail
  time:   1
  server_cmd: /home/build/buildbot/install/ofi/ofi_rhel7-debug/18/bin/fi_ep_test -p "udp"
  server_stdout: |
    fi_getinfo(): unit/ep_test.c:396, ret=-61 (No data available)

name:   fi_ep_test -p "shm"
  result: Fail
  time:   1
  server_cmd: /home/build/buildbot/install/ofi/ofi_rhel7-debug/18/bin/fi_ep_test -p "shm"
  server_stdout: |
    fi_getinfo(): unit/ep_test.c:396, ret=-61 (No data available)

name:   fi_ep_test -p "verbs"
  result: Fail
  time:   3
  server_cmd: /home/build/buildbot/install/ofi/ofi_rhel7-debug/18/bin/fi_ep_test -p "verbs"
  server_stdout: |
    Testing EPs on fabric IB-0xfe80000000000000
    Running ep_open_close [Test open and close EPs]...PASS!
    Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[32768] Opened[253]
    FAIL: fi_endpoint failed when opening new EP: ret=24 (Too many open files)
    Running ep_rand_access [Test opening and closing EPs randomly]...PASS!
    Running ep_open_sizes [Test opening and closing EPs with different context sizes]...PASS!
    Running ep_check_flags [Test get and set EP operation flags]...set tx flag 134219776 failed!
    skipped because: fi_endpoint failed when opening new EP: ret=61 (No data available)
    
    Summary: 1 tests failed
```

rajachan commented:
Verbs seems to be the only real failure there, also tripping on the max EP count check:
```
    Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[32768] Opened[253]
```

shefty commented:
Those were all separate 'tests' in the Intel CI, which reported failures.  It looks like something is not catching ENODATA as a non-failure, or we need some sort of exclude test for those.  I'll look at the test itself and see why other providers aren't supported.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+	opts.options |= FT_OPT_SIZE;
+	opts.options |= FT_OPT_SKIP_REG_MR;
+	hints->ep_attr->type = FI_EP_RDM;
+	num_eps = 50;
```
It would be simpler to init this where it is declared.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+	opts.options |= FT_OPT_SKIP_REG_MR;
+	hints->ep_attr->type = FI_EP_RDM;
```
Do we need to set a default EP type?

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+		case 'n':
+			num_eps = atoi(optarg);
+			break;
+		case 's':
```
Can we pick an unused option, since 's' is used by other tests for another purpose?

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+		}
+	}
+
+	hints->domain_attr->mr_mode = FI_MR_LOCAL | OFI_MR_BASIC_MAP;
```
INIT_OPTS should set this.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+static unsigned int num_eps, max_eps;
+struct fid_ep **ep_arr;
+struct fid_av **av_arr;
+struct fid_cq **txcq_arr, **rxcq_arr;
```
Please define a structure with the ep, av, cq fields.  Then define a single array of that structure.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+	cq_attr.wait_obj = wait_obj;
+
+	return fi_cq_open(domain, &cq_attr, cq, NULL);
+}
```
Can ft_alloc_ep_res() be used instead of the above 2 calls?  That handles additional test options.  The call can be modified to take the av, cq, cntrs, etc. as input.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,444 @@
+	FT_EP_BIND(*ep, *av, 0);
+	FT_EP_BIND(*ep, *txcq, FI_TRANSMIT);
+	FT_EP_BIND(*ep, *rxcq, FI_RECV);
+	ret = fi_enable(*ep);
```
The above code could probably call ft_enable_ep().  For MSG EP, we also need an EQ.

jswaro commented:
Ignore the Cray CI failure.

jswaro commented:
Actually... 
```
[2019-10-10T18:35:10.413Z] - name:   fi_ep_test -p "verbs;ofi_rxm"
[2019-10-10T18:35:10.414Z]   result: Fail
[2019-10-10T18:35:10.414Z]   time:   2
[2019-10-10T18:35:10.414Z]   server_cmd: /scratch/jenkins/jenkins-root/workspace/ibfabric_OFIWG_Libfabric_PR-5350/installs/fabtests/bin/fi_ep_test -p "verbs;ofi_rxm"
[2019-10-10T18:35:10.414Z]   server_stdout: |
[2019-10-10T18:35:10.414Z]     Testing EPs on fabric IB-0xfe80000000000000
[2019-10-10T18:35:10.414Z]     Running ep_open_close [Test open and close EPs]...PASS!
[2019-10-10T18:35:10.414Z]     Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[32768] Opened[253]
[2019-10-10T18:35:10.414Z]     FAIL: fi_endpoint failed when opening new EP: ret=24 (Too many open files)
[2019-10-10T18:35:10.414Z]     Running ep_rand_access [Test opening and closing EPs randomly]...PASS!
[2019-10-10T18:35:10.414Z]     Running ep_open_sizes [Test opening and closing EPs with different context sizes]...PASS!
[2019-10-10T18:35:10.414Z]     Running ep_check_flags [Test get and set EP operation flags]...set tx flag 134219776 failed!
[2019-10-10T18:35:10.414Z]     skipped because: fi_endpoint failed when opening new EP: ret=61 (No data available)
[2019-10-10T18:35:10.414Z]     
[2019-10-10T18:35:10.414Z]     Summary: 1 tests failed
```

wckzhang commented:
As far as I can tell, verbs has a bug where it cannot open its advertised number of EP's.

shefty commented:
There is no guarantee that a user can open the 'max_ep' value reported by fi_info.  There are other limiting factors, and HW resources may be shared between different object types (such as QPs, PDs, CQs, etc.).

jswaro commented:
> As far as I can tell, verbs has a bug where it cannot open its advertised number of EP's.

This has been a problem for a long time. It's related to the ulimit restrictions for a given user on a system w.r.t. the number of open file descriptors. By default, most Linux systems set this limit at ~1K FDs, but it can be set higher. So this test fails at 253 EPs, which is what I've seen here at Cray with other tests.

rajachan commented:
@jswaro I see. From a pure API stand point, applications have no way to tell though. It might make more sense to either limit the ep_cnt dynamically based on necessary rlimits/hw limits or assume the fd limit configuration is a prerequisite for users and configure CI accordingly to hit the intended max EP to catch actual issues. What do you think?

shefty commented:
To be honest, I've never liked having these 'max' values.  They invite misuse by apps.  An app cannot assume that the max of anything is ever available.  That assumes no other apps are using the resources and a specific HW limitation, where HW resources cannot be shared for different objects.

Asking providers to dynamically calculate what might be available to the current app under ideal conditions is too much, specifically when the only requirement for such a feature is coming from CI.

jswaro commented:
> @jswaro I see. From a pure API stand point, applications have no way to tell though. It might make more sense to either limit the ep_cnt dynamically based on necessary rlimits/hw limits or assume the fd limit configuration is a prerequisite for users and configure CI accordingly to hit the intended max EP to catch actual issues. What do you think?

I don't think there is a 'good' way to address this. The max values have little meaning to an application without something concrete to tie it to. Right now, the providers advertise 'best estimates' on what 'could' be provided. Personally, I see little value in those variables, but they provide an upper limit on what an application could expect, but may not see in practice. In practice, trying to derive a value based on system information is going to be faulty. If my application allocates 200 FDs for something other than libfabric, then I'd fail the test if my provider expected to be able to allocate up some dynamic value that hadn't accounted for the application's use of file descriptors. 

I think the test should likely try to allocate as many EPs as possible, and as a success criteria, return the appropriate error when the resources have been exhausted OR if it can allocate up to max. This would cover both cases where an application can actually allocate the maximum, or when it cannot due to system resource limitations.

rajachan commented:
Addressed Sean's comments a while back but forgot to push up an update, which I just did. I've also changed the `ep_open_max_simultaneously` so it doesn't fail if it doesn't hit the max ep_cnt. I've also fixed the getinfo call and ep type hints so it can run with more providers. `shm` is skipping all tests and UDP is hitting an off-by-one (opening more than it advertised as max):

```
$ ./fabtests/install/bin/fi_ep_test -p shm
Testing EPs on fabric shm
Running ep_open_close [Test open and close EPs]... Failed at 0
skipped because:
Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[1024] Opened[0]
skipped because:
Running ep_rand_access [Test opening and closing EPs randomly]...seed = 1572286481
skipped because:
Running ep_open_sizes [Test opening and closing EPs with different context sizes]...open_ep failed on size tx[1]rx[1]skipped because:
Running ep_check_flags [Test get and set EP operation flags]...skipped because:

Summary: all tests passed
```

```
$ ./fabtests/install/bin/fi_ep_test -p udp
Testing EPs on fabric UDP-IP
Running ep_open_close [Test open and close EPs]...PASS!
Running ep_open_max_simultaneously [Test opening maximum number of EPs at the same time]...Expected[128] Opened[129]
FAIL:
Running ep_rand_access [Test opening and closing EPs randomly]...PASS!
Running ep_open_sizes [Test opening and closing EPs with different context sizes]...PASS!
Running ep_check_flags [Test get and set EP operation flags]...set tx flag FI_TRANSMIT | FI_TRANSMIT_COMPLETE failed!
skipped because: fi_endpoint failed when opening new EP: ret=61 (No data available)

Summary: 1 tests failed
```

aingerson commented:
@rajachan I debugged the shm issues. Two things came up:
1) Since the provider exchanges addresses using shared memory, it hard codes the number of peers to exchange and that is 256 but it was setting ep_cnt to 1024. This was causing the issue with too many endpoints. We shouldn't need more than 256 anyways so I'll change that.
2) The utility queue implementation rounds up the sizes while the shared memory region wasn't, causing on issue is tx/rx count is not a power of 2 (the rest of fabtests uses 1024 so it was never a problem). Fixing this as well

aingerson commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,424 @@
+{
+	struct fi_av_attr av_attr;
+
+	memset(&cq_attr, 0, sizeof(av_attr));
```
Should be av_attr

aingerson commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,424 @@
+	int i, ret;
+	int testret = PASS;
+
+	for (i = 0; i < max_eps + 1; ++i) {
```
Should be max_eps, not max_eps + 1 -- this should fix the off by one error

rajachan commented:
I let the ball drop on this one. Pushed a new revision addressing all the comments.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,424 @@
+	};
+
+	max_size = MAX(max_tx_size, max_rx_size);
+	for (size = 1; size <= max_size; size += EP_TEST_SIZES_STRIDE) {
```
The 'max' sizes can be 64k or higher.  Walking through that many sizes 101 at a time will be time consuming.  We can use the existing test_sizes array here instead.  Because that array consists of values that are (2^n) and (2^n + 2 (n-1)), we can test each value as well as each value with a fixed value added to it in order to stress unlikely values.

shefty commented on fabtests/unit/ep_test.c:
```diff
@@ -0,0 +1,424 @@
+			printf("Expected open_ep to fail on sizes tx[%zu]rx[%zu]. max size: tx[%zu]rx[%zu]\n",
+			       fi->tx_attr->size, fi->rx_attr->size,
+			       max_tx_size, max_rx_size);
+			goto out;
```
Also, ret = 0 here, so I believe the test is going to pass anyway.

rajachan commented:
Will rethink the tests that use the ep/tx/rx sizes as maximums and reopen this PR when ready.
